rust/src/crypto.rs

```diff
@@ -0,0 +1,265 @@
+    (PublicKey(pk), SecretKey(sk))
+}
+
+/// A seed that can be used for key pair generation. When `KeySeed`
+/// goes out of scope, its contents will be zeroed out.
+#[derive(AsRef, AsMut, From, Serialize, Deserialize, Eq, PartialEq, Clone)]
+pub struct KeySeed(box_::Seed);
+
+impl ByteObject for KeySeed {
+    fn from_slice(bytes: &[u8]) -> Option<Self> {
+        box_::Seed::from_slice(bytes).map(Self)
+    }
+
+    fn null() -> Self {
+        Self(box_::Seed([0; box_::PUBLICKEYBYTES]))
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+/// A seed that can be used for signing key pair generation. When
+/// `KeySeed` goes out of scope, its contents will be zeroed out.
+#[derive(AsRef, AsMut, From, Serialize, Deserialize, Eq, PartialEq, Clone)]
+pub struct SignKeySeed(sign::Seed);
+
+/// Generate a new random signing key pair
+pub fn generate_sign_key_pair() -> (SignPublicKey, SignSecretKey) {
+    let (pk, sk) = sign::gen_keypair();
+    (SignPublicKey(pk), SignSecretKey(sk))
+}
+
+/// Deterministically derive a new signing key pair from the given
+/// seed
+pub fn derive_sign_key_pair_from_seed(seed: &SignKeySeed) -> (SignPublicKey, SignSecretKey) {
+    let (pk, sk) = sign::keypair_from_seed(seed.as_ref());
+    (SignPublicKey(pk), SignSecretKey(sk))
+}
+
+impl ByteObject for SignKeySeed {
+    fn from_slice(bytes: &[u8]) -> Option<Self> {
+        sign::Seed::from_slice(bytes).map(Self)
+    }
+
+    fn null() -> Self {
+        Self(sign::Seed([0; sign::PUBLICKEYBYTES]))
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+/// Public key for asymmetric authenticated encryption
+#[derive(
+    AsRef, AsMut, From, Serialize, Deserialize, Hash, Eq, Ord, PartialEq, Copy, Clone, PartialOrd,
+)]
+pub struct PublicKey(box_::PublicKey);
+
+impl ByteObject for PublicKey {
+    fn null() -> Self {
+        Self(box_::PublicKey([0_u8; box_::PUBLICKEYBYTES]))
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+
+    fn from_slice(bytes: &[u8]) -> Option<Self> {
+        box_::PublicKey::from_slice(bytes).map(Self)
+    }
+}
+
+pub const SEALBYTES: usize = sealedbox::SEALBYTES;
+
+/// Encrypt a message `m` for a recipient whose public key is
+/// `recipient_pk`. It returns the ciphertext whose length is
+/// [`SEALBYTES`] + m.len().
+///
+/// The function creates a new key pair for each message, and
+/// attaches the public key to the ciphertext. The secret key is
+/// overwritten and is not accessible after this function returns.
+pub fn encrypt(m: &[u8], recipient_pk: &PublicKey) -> Vec<u8> {
+    sealedbox::seal(m, recipient_pk.as_ref())
+}
+
+/// Decrypt the ciphertext `c` using the key pair `(pk, sk)` and
+/// returns the decrypted message.
+///
+/// If decryption fails `Err(())` is returned.
+pub fn decrypt(c: &[u8], pk: &PublicKey, sk: &SecretKey) -> Result<Vec<u8>, ()> {
+    sealedbox::open(c, pk.as_ref(), sk.as_ref())
+}
```

Comment:
we could tie encryption and decryption to the respective keys to make these calls similar to the signing. sth along the lines of
```
impl PublicKey {
    pub fn encrypt(&self, m: &[u8]) -> Vec<u8> {
        sealedbox::seal(m, self.as_ref())
    }
}

impl SecretKey {
    pub fn decrypt(&self, c: &[u8], pk: &PublicKey) -> Result<Vec<u8>, ()> {
        sealedbox::open(c, pk.as_ref(), self.as_ref())
    }
}
```
the second one could even have a shorter function signature like
```
pub fn decrypt(&self, c: &[u8]) -> Result<Vec<u8>, ()> {
    sealedbox::open(c, &self.as_ref().public_key(), self.as_ref())
}
```
but be aware that the call to `public_key()` on a secret key actually recomputes the public key from the secret key which involves a multiplication on the elliptic curve. this can become a limitation if we need to do this very often.

Reply:
I initially did it that way but it was unclear whether I should implement `decrypt` on the public or the secret key. It felt quite arbitrary to implement it on the secret key. Wdyt?

Reply:
it's not arbitrary at all. encryption is always done with the public key. decryption is always done with the secret key.
actually i'm not quite sure why the public key is even provided to the decrypt function in libsodium. maybe some internal consistency checks or just a relic. but regarding the mathematics, it is not needed there.